README
..	-*- mode: rst-mode -*-
..
.. Version number is filled in automatically.
.. |version| replace:: 1.62-2

===============================================
Ruby Bindings for Broccoli
===============================================

.. rst-class:: opening

    This is the broccoli-ruby extension for Ruby which provides access
    to the Broccoli API.  Broccoli is a library for
    communicating with the Bro Intrusion Detection System.


Download
========

You can find the latest Broccoli-Ruby release for download at
https://www.zeek.org/download.

Broccoli-Ruby's git repository is located at
https://github.com/zeek/broccoli-ruby

This document describes Broccoli-Ruby |version|. See the ``CHANGES``
file for version history.


Installation
============

To install the extension:

1. Make sure that the ``broccoli-config`` binary is in your path.
   (``export PATH=/usr/local/bro/bin:$PATH``)

2. Run ``sudo ruby setup.rb``.

To install the extension as a gem (suggested):

1. Install `rubygems <http://rubygems.org>`_.

2. Make sure that the ``broccoli-config`` binary is in your path.
   (``export PATH=/usr/local/bro/bin:$PATH``)

3. Run, ``sudo gem install rbroccoli``.

Usage
=====

There aren't really any useful docs yet.  Your best bet currently is
to read through the examples.

One thing I should mention however is that I haven't done any optimization
yet.  You may find that if you write code that is going to be sending or
receiving extremely large numbers of events, that it won't run fast enough and
will begin to fall behind the Bro server.  The dns_requests.rb example is
a good performance test if your Bro server is sitting on a network with many
dns lookups.

Contact
=======

If you have a question/comment/patch, see the Bro `contact page
<http://www.bro.org/contact/index.html>`_.


CHANGES

1.62-2 | 2018-12-07 16:28:24 -0600

  * Update github/download link (Jon Siwek, Corelight)

  * Update submodules to use github.com/zeek (Jon Siwek, Corelight)

1.62 | 2018-08-31 15:25:31 -0500

  * Improve `make dist` (Jon Siwek, Corelight)

1.61-11 | 2018-08-06 16:01:46 -0500

  * Fix deprecation warning for swig_add_module (Jon Siwek, Corelight)

1.61-5 | 2018-05-15 15:39:06 +0000

  * Updating submodule.

1.61 | 2017-05-26 08:27:27 -0500

 * Release 1.61.

1.60 | 2016-10-27 14:42:49 -0700

 * Release 1.60.

1.59 | 2016-06-30 16:55:31 -0700

  * Add README.rst -> README symlink. (Johanna Amann)

1.58 | 2015-04-27 08:25:18 -0700

  * Release 1.58

1.57 | 2014-04-03 15:53:50 -0700

  * Release 1.57

1.56 | 2013-10-14 09:24:55 -0700

  * Updating copyright notice. (Robin Sommer)

  * Fix for setting REPO in Makefile. (Robin Sommer)

1.55 | 2013-09-23 14:42:03 -0500

  * Update 'make dist' target. (Jon Siwek)

  * Change submodules to fixed URL. (Jon Siwek)

  * Switching to relative submodule paths. (Robin Sommer)

  * s/bro-ids.org/bro.org/g. (Robin Sommer)

1.54 | 2012-08-01 13:56:22 -0500

  * Fix configure script to exit with non-zero status on error (Jon Siwek)

1.53 | 2012-06-11 17:25:05 -0700

  * Update bindings to work with Broccoli's IPv4-mapped BroAddrs. (Jon Siwek)

  * Fix count/enum being treated same as addr. (Jon Siwek)

  * Update broccoli-ruby for IPv6 addr/subnet support. Addresses #448.
    (Jon Siwek)

  * Raise minimum required CMake version to 2.6.3 (Jon Siwek)

1.52 | 2012-01-09 16:11:01 -0800

  * Submodule README conformity changes (Jon Siwek)

1.51-10 | 2011-11-07 05:44:17 -0800

  * Ignoring some SWIG warnings. Addresses #388. (Jon Siwek)

  * Changes to broccoli-ruby installation scheme. Fixes #652.

    - `--home` and `--prefix` configure options are now respected when
      installing as the main CMake project.  If not given, the Ruby

    - When being installed as a CMake sub-project, then the
      "home"-style installation is performed. (Jon Siwek)

1.51 | 2011-10-26 13:51:22 -0700

  * Compile SWIG bindings with no-strict-aliasing (addresses #644).
    (Jon Siwek)

1.5 | 2011-10-25 17:41:31 -0700

  * Make dist now cleans the copied source. (Jon Siwek)

  * Add configure-time check that swig can generate Ruby wrappers.
    Addresses #642. (Jon Siwek)

  * Distribution cleanup. (Robin Sommer)

  * Updates to work with communication API changes. (Seth Hall)

  * Reorganized the module names.  From ruby, a user now loads the
    "broccoli" module.  This automatically pulls in the swig wrapper
    named "broccoli_ext".  (Seth Hall)

  * Building with cmake completely works now. (Seth Hall)

  * Updates for the change to 64-bit ints. (Seth Hall)

  * Fixes for the example script. (Seth Hall)

  * New example script that points out a bug in broccoli. (Seth Hall)

  * Remove the 'net' type from Broccoli ruby bindings. Addresses #535.
    (Jon Siwek)

  * Install binaries with an RPATH (Jon Siwek)

1.4 | 2011-02-25 21:26:49 -0500

  * Cleaning up and adding a configure script. (Seth Hall)

  * Ruby 1.8 is now required. (Seth Hall)

  * CMake fixes. (Seth Hall and Jon Siwek)

  * Initial import. (Seth Hall)




lib/bro.rb
require 'broccoli'
require 'time'

require 'Bro/connection'
require 'Bro/event'
require 'Bro/record'

module Bro
  include Broccoli
  
  TYPES = {:unknown => BRO_TYPE_UNKNOWN, # not really sure how this should be handled.
           :bool => BRO_TYPE_BOOL,
           :int => BRO_TYPE_INT,
           :count => BRO_TYPE_COUNT,
           :counter => BRO_TYPE_COUNTER,
           :double => BRO_TYPE_DOUBLE,
           :time => BRO_TYPE_TIME,
           :interval => BRO_TYPE_INTERVAL,
           :string => BRO_TYPE_STRING,
           :enum => BRO_TYPE_ENUM,
           :timer => BRO_TYPE_TIMER,
           :port => BRO_TYPE_PORT,
           :addr => BRO_TYPE_IPADDR,
           :subnet => BRO_TYPE_SUBNET,
           :record => BRO_TYPE_RECORD,
           # These are not handled by the ruby binding.
           :packet => BRO_TYPE_PACKET,
           :max => BRO_TYPE_MAX,
           # All types below are NOT handled by broccoli.
           :pattern => BRO_TYPE_PATTERN,
           :any => BRO_TYPE_ANY,
           :table => BRO_TYPE_TABLE,
           :union => BRO_TYPE_UNION,
           :list => BRO_TYPE_LIST,
           :func => BRO_TYPE_FUNC,
           :file => BRO_TYPE_FILE,
           :vector => BRO_TYPE_VECTOR,
           :type => BRO_TYPE_TYPE,
           :error => BRO_TYPE_ERROR
          }
  
  def Bro.current_time_f
    Broccoli::bro_util_current_time
  end

  def Bro.current_time
    Time.at( current_time_f() )
  end

  def Bro.debug_calltrace=(v)
    Broccoli::bro_debug_calltrace=v
  end

  def Bro.debug_messages=(v)
    Broccoli::bro_debug_messages=v
  end
end

class Broccoli::BroPort
  @@protocols = {0=>'ip', 1=>'icmp', 2=>'igmp', 3=>'ggp', 4=>'ipv4',
                 6=>'tcp', 7=>'st', 8=>'egp', 9=>'pigp', 10=>'rccmon',
                 11=>'nvpii', 12=>'pup', 13=>'argus', 14=>'emcon',
                 15=>'xnet', 16=>'chaos', 17=>'udp', 18=>'mux', 19=>'meas',
                 20=>'hmp', 21=>'prm', 22=>'idp', 23=>'trunk1', 24=>'trunk2',
                 25=>'leaf1', 26=>'leaf2', 27=>'rdp', 28=>'irtp', 29=>'tp',
                 30=>'blt', 31=>'nsp', 32=>'inp', 33=>'sep', 34=>'3pc',
                 35=>'idpr', 36=>'xtp', 37=>'ddp', 38=>'cmtp', 39=>'tpxx',
                 40=>'il', 41=>'ipv6', 42=>'sdrp', 43=>'routing', 
                 44=>'fragment', 45=>'idrp', 46=>'rsvp', 47=>'gre', 48=>'mhrp',
                 49=>'bha', 50=>'esp', 51=>'ah', 52=>'inlsp', 53=>'swipe', 
                 54=>'nhrp', 58=>'icmpv6', 59=>'nonext', 60=>'dstopts',
                 61=>'ahip', 62=>'cftp', 63=>'hello', 64=>'satexpak', 
                 65=>'kryptolan', 66=>'rvd', 67=>'ippc', 68=>'adfs', 
                 69=>'satmon', 70=>'visa', 71=>'ipcv', 72=>'cpnx', 73=>'cphb',
                 74=>'wsn', 75=>'pvp', 76=>'brsatmon', 77=>'nd', 78=>'wbmon',
                 79=>'wbexpak', 80=>'eon', 81=>'vmtp', 82=>'svmtp', 
                 83=>'vines', 84=>'ttp', 85=>'igp', 86=>'dgp', 87=>'tcf', 
                 88=>'igrp', 89=>'ospfigp', 90=>'srpc', 91=>'larp', 92=>'mtp',
                 93=>'ax25', 94=>'ipeip', 95=>'micp', 96=>'sccsp', 
                 97=>'etherip', 98=>'encap', 99=>'apes', 100=>'gmtp', 
                 103=>'pim', 108=>'ipcomp', 113=>'pgm', 254=>'divert', 
                 255=>'raw'}
  def to_s
    #"#{port_num}/#{@@protocols[port_proto]}"
    port_num.to_s
  end
end
